Add dynamic properties to WebACS

sabicalija opened the pull request.

sabicalija committed:
Add dynamic properties to WebACS

klues commented on view/propertyEditor.js:
```diff
@@ -245,77 +244,118 @@ ACS.propertyEditor = function (modelList, // ACS.modelList
+			if (isDynProp) {
+
+				log.debug('downloading (dynamic) properties list');
+				setBaseURI(ACS.areBaseURI + '/rest/');
```
why this is needed here? shouldn't this be done once somewhere at setup?

deinhofer replied:
In the C# ACS the dyn properties are only fetched, if the ACS is connected to the ARE and the currently edited model is also deployed to the ARE.
-->
You should only fetch the dyn properties, if the ARE Status is
![image](https://user-images.githubusercontent.com/4621810/54362985-47683480-466a-11e9-8d5e-9fec08422526.png)
In all other cases please show a normal text input field filled out with the current value of the field.

sabicalija replied:
Dynamic properties are now loading only when ARE and WebACS are synchronised. 

In all other cases, only an input field is displayed for dynamic properties (with the current value, see https://github.com/asterics/WebACS/pull/36/commits/e4592daca1b0c20471cccd3c5302e91ab763e84a)

deinhofer commented on view/propertyEditor.js:
```diff
@@ -245,77 +244,118 @@ ACS.propertyEditor = function (modelList, // ACS.modelList
+				function fetchDynProperties(cellToAdd, elementId, currentValue, id, key) {
+					getRuntimeComponentPropertyList(function(data, httpStatus) {
+						var entries = data ? JSON.parse(data) : null;
+						entries.unshift("");
```
this will produce an error, if entries == null

what about
```js
var entries = data ? JSON.parse(data) : new Array("");
```
if you need an array with one entry with value "".

klues replied:
yes, makes sense, simpler just `[""]` instead of `new Array("")`

klues replied:
no, sorry - better to move line `entries.unshift("");` in the else-statement in line `269`. @deinhofer's solution breaks the if-check on line `263`, since then `entries.length` will be `1`

sabicalija replied:
Like this https://github.com/asterics/WebACS/pull/36/commits/7feaf22054cf81ffe5e344b8e290cd61161ca996?

deinhofer replied:
No, in the 'else' branch entries could still be 'null' and hence entries.unshift would produce an error.

new suggestion do all the magic like this:
```javascript
var entries = data ? JSON.parse(data).unshift("") : [""];
```

Then you have an extra empty field in any case.
Then you have to change the if-condition to check if entries.length === 1

klues replied:
simpler: 
`var entries = data ? JSON.parse(data).unshift("") : null;`
and then 
`if(!entries)`

it makes no sense to create an empty array just to check afterwards if the length == 1.

sabicalija replied:
But the if/else should make sure that entries is not null in the else branch, shouldn't it?

klues replied:
ok, what I said is nonsense - then we would have the problem that if `JSON.parse(data)` is empty, that we do not come in the first if-statement... I should go home ^^

sabicalija replied:
just figured that what I said was nonsense too... :smile: I'll review later at leisure and push a new suggestion.

sabicalija replied:
What about following suggestion: https://github.com/asterics/WebACS/pull/36/commits/8f5d9c76a48891e049da135c1429df9d7a587eef ?

deinhofer replied:
Bingo!! :+1: loogs good!!

deinhofer replied:
> What about following suggestion: [8f5d9c7](https://github.com/asterics/WebACS/commit/8f5d9c76a48891e049da135c1429df9d7a587eef) ?

Bingo!! :+1: loogs good!!

sabicalija replied:
Still not ideal, but I'll suggest to refactor it when `webpack` and **es6** features are available (https://github.com/asterics/WebACS/pull/36/commits/f9df73106e86ea6564be58f7f755a7c642efa2fd).

Note: It is not ensured that what we're looking at is actually an array (thus having the property `length`).

sabicalija replied:
The only way to make sure that `unshift()` is called correctly - at least the only I can think of - would be to use `instanceof` (see https://github.com/asterics/WebACS/pull/36/commits/f831f10dc8c70f9907eb1ab64175c7320430363c). Wouldn't it? Like:

```js
  if (entries instanceof Array) {
     ...
  }
```
We could check if the property `length` is available too, but I think using `instanceof` is better in this case.

**EDIT**
I've refactored the code a little, to remove some duplicates (see https://github.com/asterics/WebACS/pull/36/commits/ad090d65c9d66f637e7d1666253ae6dbe1d2a037).

sabicalija replied:
In future changes, `getRuntimeComponentPropertyList()` should return a valid array or `null` otherwise. This way we could avoid dispersing this kind of logic and reduce the test to a simple `null` comparison for users of this functions. What do you think?

sabicalija replied:
Actually, the first suggestion is probably the best, and should work too... :grin:

```js
var entries = data ? JSON.parse(data) : [];
if (entries.length === 0) { 
   // input
} else {
   // selection
}
```

sabicalija replied:
> yes, makes sense, simpler just `[""]` instead of `new Array("")`

Both of your suggestions create an array with a length of 1.  Simply `[]` should be possible too. Like:

```js
var entries = data ? JSON.parse(data) : [];
```

What syntax has broader browser support? `[]` or `new Array()`.

klues replied:
both should be supported in all browsers, however I vote for `[]` since it is shorter, simpler and as it seems also faster: https://stackoverflow.com/questions/7375120/why-is-arr-faster-than-arr-new-array

sabicalija replied:
I like the shorter form too ... generally :smirk:

sabicalija committed:
Load dynamic properties only when model in ARE and ACS are n'sync :smirk:.

sabicalija committed:
Correct empty field creation for dynamic properties

sabicalija committed:
Correct processing of received dyn. property data

sabicalija committed:
Remove unwanted checks

sabicalija committed:
Add instanceof test for received entries

sabicalija committed:
Remove duplicates.

klues commented on view/propertyEditor.js:
```diff
@@ -245,77 +244,143 @@ ACS.propertyEditor = function (modelList, // ACS.modelList
+						getRuntimeComponentPropertyList(function(data, httpStatus) {
+							var entries = data ? JSON.parse(data) : null;
+							if (entries === null || !(entries instanceof Array) || entries.length === 0) {
+								if (entries === null) console.warn('cannot process received data');
```
so, new day - new try to review :D
(1) `(entries === null || !(entries instanceof Array) || entries.length === 0)` could be simplified to
 `(!(entries instanceof Array) || entries.length === 0)` since `(null instanceof Array) == false`

(2) `if (entries === null) console.warn('cannot process received data');`
is semantically not correct, because if the received data could not be processed `JSON.parse(data)` above would throw an error and this line would not be reached. If you want to catch JSON parsing errors you would have to  surround `JSON.parse()` with try/catch. However I don't know if this is necessary - depends how much we trust `ARE` to deliver correct data.

(3) I don't like if/else statements without curly braces.

If you are tired of changing these lines of code over and over again, for me it would be also OK to merge as-is, since functional it should be correct now.
Since

sabicalija replied:
I'm not tired of changing those lines :relaxed: Don't worry. I'll integrate (1) and (3). For (2) I'll change the text to "received null" (??). Or should I drop it entirely?

deinhofer replied:
new day, new insights!
I think we all focused onto the wrong line!!

actually ```data``` can be null and this leads to a filled entries array or null.
So why not simply checking data in the if condition and to the rest within the if-body?

```javascript
getRuntimeComponentPropertyList(function(data, httpStatus) {
    if(data === null) {
        // insert empty field
    } else {
       entries=JSON.parse(data);
       //insert dropbox
   }
```

klues replied:
@deinhofer: the problem with your solution is that `JSON.parse(data)` can also be an empty array and in this case we also want to insert the empty field.

@sabicalija: for (2) I would just simply do `console.warn('no dynamic properties from ARE - using textfield instead of combobox')` and not distinguish between the two cases.

Maybe the cleanest solution would be a separate function:
```
function parseEntries (jsonString) {
    try{
        let parsed = JSON.parse(jsonString);
        return parsed instanceof Array ? parsed : [];
    } catch(error) {}
    return [];
}
```

and then just:
```
getRuntimeComponentPropertyList(function(data, httpStatus) {
    let entries = parseEntries(data);
    if(entries.length === 0) {
        // insert empty field
    } else {
       //add empty element to entries
       //insert dropbox
   }
```

klues replied:
the function `parseEntries` could also be generalized to something like `parseJSONtoArray()` and then added to some util class, if any is available.

sabicalija replied:
Proposed changes are now added (see https://github.com/asterics/WebACS/pull/36/commits/52a208603f7c636d134dfe23eebc81371ce12633 and https://github.com/asterics/WebACS/pull/36/commits/6e26a7ba0b33ef0fb04beb9d2f6de0a27cd9fa74). Should be fine for now. However, I'd like to use a generic `parseJSONtoArray()` in future.

sabicalija committed:
Refactor and modularize dynamic list properties parsing

sabicalija committed:
Fix typo

klues commented on view/propertyEditor.js:
```diff
@@ -245,77 +244,141 @@ ACS.propertyEditor = function (modelList, // ACS.modelList
+						console.warn('no dynamic properties from ARE - using textfield instead of combobox');
+						var inputElement = document.createElement('input');
+						inputElement.value = currentValue;
+						inputElement.setAttribute("id", elementId);
```
Great - looks quite clean now 👍 
one thing I just noticed: in case of the plain input element there is no `addEventListener()` call like in the case with the combobox. I don't know if it is needed?! Beside of this I think we finally can merge 😃

sabicalija replied:
Should probably be in there. Maybe I've missed it or I've rationalized during the changes of this pr. :smile: Anyway, I'll add it. What is strange is that during testing, I've changed the value of this input element (with disconnected ARE), stored the model and loaded it afterwards. And it worked.

I've apparently also overlooked the outermost else branch at lines 305-308. Should call `insertInputElement()` there too. Shouldn't I?

klues replied:
> I've changed the value of this input element (with disconnected ARE), stored the model and loaded it afterwards. And it worked.

I think the `change` event listener is for live updates of the values, if connected with ARE and the model is running on ARE. For storing/loading the model it should not make a difference.

> ... at lines 305-308. Should call insertInputElement() there too. Shouldn't I?

yes, makes sense ;)

deinhofer replied:
It does not seem to work any more.
I did the following steps:
1. create a model with Knx plugin
2. upload model
3. start model
4. select plugin

There is no combobox and there are error messages in the console:
![image](https://user-images.githubusercontent.com/4621810/55815374-8ca84680-5af0-11e9-959a-dac75900fe78.png)

sabicalija replied:
I just tried the 4 steps you mention. I cannot reproduce the described error. My model is, however, the most simple case - just the KNX plugin :smile:. Have you used a more sophisticated model?

![image](https://user-images.githubusercontent.com/13609768/55828863-a4da8e80-5b0d-11e9-9645-b7903470d478.png)

What I did was the following:
* place KNX plugin
* upload model
* start model
* select combo box (dyn. prop.)

Can you show me your model tomorrow? Or maybe add it to your comment - but I'm not sure if you can add attachments to post after all.

deinhofer replied:
ok, sorry my mistake!!
I forgot to start the corresponding ARE with the new REST call, see https://github.com/asterics/AsTeRICS/pull/308

Works like a charme :+1: 
Let's merge!

deinhofer committed:
updated areCommunicator.js to https://github.com/asterics/AsTeRICS/blob/b17cfb7ecf8ba9d7bcd22706fc3e87e1e57e86da/ARE_RestAPIlibraries/clientExample/javascript/areCommunicator.js
 * There were many newer functions missing
 * now use renamed method getRuntimeComponentPropertyDynamic in view/propertyEditor.js

used formatting of Visual Studio Code which changed a lot, maybe we should fix that again.

sabicalija reviewed (commented):
I didn't review everything thoroughly, since the changes are mainly due to reformatting of VSCode. (Almost?) All the missing function of `areCommunicator.js` have been added with `@asterics/rest`.

I don't think this additions were necessary for this PR. Are we going to maintain `areCommunicator.js` in the `WebACS` repository, the `AsTeRICS` repository and in a future repository for the `npm` package "@asterics/rest"?
